pkg/microservice/policy/core/service/policy.go

```diff
@@ -0,0 +1,211 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package service
+
+import (
+	"fmt"
+
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/util/sets"
+
+	"github.com/koderover/zadig/pkg/microservice/aslan/core/label/service"
+	"github.com/koderover/zadig/pkg/shared/client/label"
+
+	"github.com/koderover/zadig/pkg/microservice/policy/core/repository/models"
+	"github.com/koderover/zadig/pkg/microservice/policy/core/repository/mongodb"
+)
+
+type Policy struct {
+	Name        string  `json:"name"`
+	Description string  `json:"description"`
+	UpdateTime  int64   `json:"update_time"`
+	Rules       []*Rule `json:"rules,omitempty"`
+}
+
+type ListPolicyResp struct {
+	Policies []Policy `json:"policies"`
+}
+
+func CreatePolicy(ns string, policy *Policy, _ *zap.SugaredLogger) error {
+	obj := &models.Policy{
+		Name:      policy.Name,
+		Namespace: ns,
+	}
+
+	for _, r := range policy.Rules {
+		obj.Rules = append(obj.Rules, &models.Rule{
+			Verbs:           r.Verbs,
+			Kind:            r.Kind,
+			Resources:       r.Resources,
+			MatchAttributes: r.MatchAttributes,
+		})
+	}
+	return mongodb.NewPolicyColl().Create(obj)
+}
+
+func CreatePolicies(ns string, policies []*Policy, _ *zap.SugaredLogger) error {
+	var objs []*models.Policy
+	for _, policy := range policies {
+		obj := &models.Policy{
+			Name:      policy.Name,
+			Namespace: ns,
+		}
+
+		for _, r := range policy.Rules {
+			obj.Rules = append(obj.Rules, &models.Rule{
+				Verbs:           r.Verbs,
+				Kind:            r.Kind,
+				Resources:       r.Resources,
+				MatchAttributes: r.MatchAttributes,
+			})
+		}
+		objs = append(objs, obj)
+	}
+	return mongodb.NewPolicyColl().BulkCreate(objs)
+}
+
+func UpdatePolicy(ns string, policy *Policy, _ *zap.SugaredLogger) error {
+	obj := &models.Policy{
+		Name:      policy.Name,
+		Namespace: ns,
+	}
+
+	for _, r := range policy.Rules {
+		obj.Rules = append(obj.Rules, &models.Rule{
+			Verbs:           r.Verbs,
+			Kind:            r.Kind,
+			Resources:       r.Resources,
+			MatchAttributes: r.MatchAttributes,
+		})
+	}
+	return mongodb.NewPolicyColl().UpdatePolicy(obj)
+}
+
+func UpdateOrCreatePolicy(ns string, policy *Policy, _ *zap.SugaredLogger) error {
+	obj := &models.Policy{
+		Name:      policy.Name,
+		Namespace: ns,
+	}
+
+	for _, r := range policy.Rules {
+		obj.Rules = append(obj.Rules, &models.Rule{
+			Verbs:           r.Verbs,
+			Kind:            r.Kind,
+			Resources:       r.Resources,
+			MatchAttributes: r.MatchAttributes,
+		})
+	}
+	return mongodb.NewPolicyColl().UpdateOrCreate(obj)
+}
+
```

Comment:
write a function which converts ```*Policy``` to ```*models.Policy``` would do some help of reduce code redundancy

Reply:
do not need all fields , do not have translate layer ， we can discuss this

Reply:
it's nothing about `translate layer`, there are too many dulpicated codes in ```UpdatePolicy``` and ```UpdateOrCreatePolicy```